blivet/fcoe.py

```diff
@@ -110,27 +111,69 @@ def addSan(self, nic, dcb=False, auto_vlan=True):
         log.info("Activating FCoE SAN attached to %s, dcb: %s autovlan: %s",
                  nic, dcb, auto_vlan)
 
-        util.run_program(["ip", "link", "set", nic, "up"])
-
         rc = 0
+        out = ""
+        timeout = 60
         error_msg = ""
         if dcb:
+
             self._startLldpad()
-            util.run_program(["dcbtool", "sc", nic, "dcb", "on"])
-            util.run_program(["dcbtool", "sc", nic, "app:fcoe",
-                                                "e:1", "a:1", "w:1"])
-            rc, out = util.run_program_and_capture_output(["fipvlan", "-c", "-s", "-f",
-                                               "-fcoe", nic])
+            while timeout > 0:
+                rc, out = util.run_program_and_capture_output(
+                    ["lldptool", "-p"])
+                if not rc:
+                    break
+                log.info("Waiting for lldpad to be ready")
+                timeout -= 1
+                time.sleep(1)
+
+            while timeout > 0:
+                rc, out = util.run_program_and_capture_output(
+                    ["dcbtool", "sc", nic, "dcb", "on"])
+                if not rc:
+                    break
+                log.info("Retrying to turn dcb on")
+                timeout -= 1
+                time.sleep(1)
+
+            while timeout > 0:
+                rc, out = util.run_program_and_capture_output(
+                    ["dcbtool", "sc", nic, "pfc", "e:1", "a:1", "w:1"])
+                if not rc:
+                    break
+                log.info("Retrying to set up dcb with pfc")
+                timeout -= 1
+                time.sleep(1)
+
+            while timeout > 0:
+                rc, out = util.run_program_and_capture_output(
+                    ["dcbtool", "sc", nic, "app:fcoe", "e:1", "a:1", "w:1"])
+                if not rc:
+                    break
+                time.sleep(1)
+                log.info("Retrying to set up dcb for fcoe")
```

Comment:
Looks like this "wait until utility exits successfully or time out" could be a useful function probably in ``iutil.py`` as it may be useful for other modules too.